Do a more thorough check when a bad try_list is detected

puzzledsab opened the pull request:
First part of #2320.

If the last server in the priority list is deactivated and it was the only missing server in the article try_list then none of the active servers will check the article. This means that the article, file and nzb may never be registered as finished because that is only done after checking an article.

The bad try_list check will notice this because if there are 3 active servers and the last is deactivated then there will be 2 servers in the try_list and 2 active servers. If this is still the case after checking that all the servers in the try_list are currently active then the article will be marked as failed.

Safihre commented:
Looking good, I didn't think of this indeed!

Currently I'm less available due to (happy) family circumstances 🍼, will look at in depth in a few days.

puzzledsab commented:
Congratulations :)

I have changed the test so that it doesn't delete inactive servers. I was afraid that there might be cases where the same server would try to download the same article multiple times if it was disabled and then reenabled.

Regarding the DNS lookup issue, it turns out it was already supposed to disable the server for 10 minutes in `maybe_disable_server()`, but it didn't trigger because `get_server_addrinfo()` didn't return `False` like it expected. I fixed it by setting it to `False` explicitly because server.info can contain different false values (ie `None`) that should be treated differently. I included it in the same PR because it was such a small change.

Safihre commented on sabnzbd/nzbqueue.py:
```diff
@@ -896,11 +901,21 @@ def stop_idle_jobs(self):
             if len(nzo.try_list) >= nr_servers:
                 # Maybe the NZF's need a reset too?
                 for nzf in nzo.files:
+                    if nzo.removed_from_queue:
```
Maybe we should wrap this whole section in `with NZBQUEUE_LOCK ` so we don't have to check for these things?
Because also below I wonder if we won't end up in a race condition with nzf.articles?
If one is removed during looping then it will result in a crash.

puzzledsab replied:
I only added those checks for performance. Initially in my test it would go through all the files and articles in the nzo, even though the quick completion test would move the nzo out of the queue quickly because less than 100% including par2 was available.

My test consisted of an nzb where all the articles were missing on all the servers, and the lowest priority server had an invalid host name. I then disabled the invalid server when all the valid servers had tried all the articles. You have to comment out `self.info = False` in this PR to do the same test.

The locking problem I don't properly understand. Intuitively I would agree, but the more I dig into it, the stranger it seems. `sabnzbd.NzbQueue.register_article()` is called inside the article loop and did not cause problems in my tests. I assume it removed articles from the list while looping.

Also the downloader thread loops through the articles in an unlocked mode in `get_articles()` even though the decoder thread regularly deletes articles from `nzf.articles` through `NzbQueue.register_article()`. That doesn't seem to cause problems either. What am I missing?

I could change it so that the check is done before looping the articles in `nzf.get_articles()` in the downloader thread, using `nzf.articles[:]` to avoid deleting from the same list as it's looping. It would reduce some of the concurrency. It shouldn't cause a noticeable performance hit in normal circumstances if the servers parameter is changed so that it only includes active servers.

Safihre replied:
Since this is only called in case an idle job is found, do we really need to worry about performance? I guess it doesn't hurt, the `if`'s just bothered me a tiny bit 😅

From what I can see, the only place `nzf.articles` has articles removed is using `nzf.remove_article`. Which is called from `nzo.remove_article` which is locked by `NZO_LOCK`. 
https://github.com/sabnzbd/sabnzbd/blob/be7ae3d151339fd379d1774f0a86b688887df9ff/sabnzbd/nzbstuff.py#L1190-L1191
So I think we either need to wrap this is `NZO_LOCK` or use `nzf.articles[:]`.

puzzledsab replied:
I removed one of the if blocks. I also realized that removing elements from the list I'm looping over creates problems because the element position is shifted. If element 0 is removed then element 1 becomes 0 and so on. The loop has already read element 0 so the original element 1 will be skipped. If trying to remove every element then this causes half of them to be skipped silently if [:] isn't used.

Regarding the lock, I was thinking about this part: https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/nzbstuff.py#L432

The remove_article methods are called from other threads than the downloader. They could theoretically remove an article from nzf.articles while the downloader thread is using the get_articles() code to loop over the article list. As far as I can tell no level of the get_articles system is locked so using the NZO_LOCK before removing doesn't fix it. I don't think it would print any errors or crash if it happens but it may skip one of the remaining articles in the list. Adding [:] would avoid it.

puzzledsab replied:
Btw, `NzbQueue.register_article()` is also using `nzo.remove_article()`, so NZO_LOCK will be set while an article is removed.

Safihre commented on sabnzbd/nzbqueue.py:
```diff
@@ -896,11 +901,21 @@ def stop_idle_jobs(self):
+                            if article.all_servers_in_try_list(active_servers):
+                                sabnzbd.NzbQueue.register_article(article, success=False)
+                                nzo.increase_bad_articles_counter("missing_articles")
+                                nzf.has_bad_articles = True
```
I just removed this everywhere, was a leftover from the failed new QuickCheck.

Safihre commented:
It looks good to me. 
Do you think this fully resolves #2320?

puzzledsab commented:
Yes.

Safihre commented:
Thanks! 🎉
